File: evm/src/observer/curation/gallery.ts
import { Chain } from "@prisma/client";
import { CHAIN, CONTRACT_CURATION, EVENTOFFSET_ID } from "../../config";
import { prisma } from "../../io";
import { redis } from "../../io/redis";
import { TypedEvent } from "../../typechain/common";
import { GalleryChangedEvent } from "../../typechain/Curation";
import { State } from "../../types";
import { handleError, Observer } from "../observer";

export class GalleryObserver extends Observer {
  async processAll<T extends TypedEvent>(
    state: State,
    events: T[],
    startBlockNo: number,
    endBlockNo: number
  ): Promise<State> {
    const newEvents = events.filter(
      (e) => e.blockNumber > Number(state.gallery_excuted_offset)
    );
    return super.processAll(state, newEvents, startBlockNo, endBlockNo);
  }
  async process<T extends TypedEvent>(
    state: State,
    event: T
  ): Promise<{ success: boolean; state: State }> {
    const blockNo = BigInt(event.blockNumber);
    const [
      id,
      eventType,
      collection,
      tokenId,
      owner,
      spaceType,
      name,
      metadataURI,
      isToken,
      timestamp,
      payees,
      rates,
      commissionPool,
      admissions,
    ] = (event as GalleryChangedEvent).args;

    const commissionRates = payees.reduce((acc, key, i) => {
      return { ...acc, [key]: rates[i].toString() };
    }, {});
    const createGallery = prisma.curationGallery.upsert({
      where: {
        index_root: {
          index: id.toBigInt(),
          root: CONTRACT_CURATION,
        },
      },
      create: {
        index: id.toBigInt(),
        root: CONTRACT_CURATION,
        chain: CHAIN as Chain,
        owner,
        spaceType,
        name,
        metadataURI,
        commissionRates,
        commissionPool,
        admissions: admissions.join(","),
      },
      update: {
        owner,
        spaceType,
        name,
        metadataURI,
        commissionRates,
        commissionPool,
        admissions: admissions.join(","),
      },
    });

    const updateState = prisma.eventOffset.update({
      where: {
        id: +EVENTOFFSET_ID,
      },
      data: {
        gallery_excuted_offset: blockNo,
      },
    });
    try {
      const [gallery, updatedState] = await prisma.$transaction([
        createGallery,
        updateState,
      ]);
      const response = await fetch(metadataURI);
      const metadata = await response.json();
      const galleryData = JSON.stringify({
        ...metadata,
        index: Number(gallery.index),
        root: CONTRACT_CURATION,
        chain: CHAIN as Chain,
        owner,
        spaceType,
        name,
        metadataURI,
        commissionRates,
        commissionPool,
        admissions: admissions.join(","),
      });
      redis.set(`mixverse:curation:${metadata.id}`, galleryData);
      redis.set(
        `mixverse:curation:${gallery.root}:${Number(gallery.index)}`,
        galleryData
      );
      if (updatedState.gallery_excuted_offset != blockNo) {
        return { success: false, state };
      }
      const newState = {
        ...state,
        gallery_excuted_offset: blockNo,
      };
      return { success: true, state: newState };
    } catch (e) {
      handleError(e);
      return { success: false, state };
    }
  }
}

File: aptos/observer/curation/gallery.py
import json
from typing import List, Tuple
from observer.observer import Observer
from model.curation.gallery_event import GalleryEvent, GalleryEventData
from model.state import State
from model.event import Event
from common.db import prisma_client
from common.redis import redis_cli
from config import config
from common.util import new_uuid
from prisma import enums, Json
import requests


class GalleryEventObserver(Observer[GalleryEvent]):

    async def process_all(self, state: State, events: List[Event[GalleryEvent]]) -> State:
        return await super().process_all(state, events)

    async def process(self, state: State, event: Event[GalleryEvent]) -> Tuple[State, bool]:
        new_state = state
        seqno = event.sequence_number
        data = GalleryEventData(**event.data)
        index = int(data.id)

        async with prisma_client.tx(timeout=60000) as transaction:
            result = await transaction.curationgallery.upsert(
                where={
                    'index_root': {
                        'index': index,
                        'root': config.curation.address()
                    }
                },
                data={
                    'create': {
                        'id': new_uuid(),
                        'chain': enums.Chain.APTOS,
                        'index': index,
                        'root': config.curation.address(),
                        'name': data.name,
                        'owner': data.owner,
                        'spaceType': data.space_type,
                        'metadataUri': data.metadata_uri,
                        'commissionRates': Json(dict(zip(data.payees, data.commission_rates))),
                        'commissionPool': ""
                    },
                    'update': {
                        'chain': enums.Chain.APTOS,
                        'name': data.name,
                        'owner': data.owner,
                        'spaceType': data.space_type,
                        'metadataUri': data.metadata_uri,
                        'commissionRates': Json(dict(zip(data.payees, data.commission_rates)))
                    }
                }
            )
            if result == None:
                raise Exception(
                    f'[Curator gallery]: Failed to update gallery({data})')
            r = requests.get(data.metadata_uri)
            metadata = r.json()
            curation_id = metadata['id']
            metadata = metadata.update(result)
            metadata_str = json.dumps(
                metadata, indent=4, sort_keys=True, default=str)
            key_curation_id = f'mixverse:curation:{curation_id}'
            key_curation_root_index = f"mixverse:curation:{config.curation.address()}:{index}"
            redis_cli.set(key_curation_id, metadata_str)
            redis_cli.set(key_curation_root_index, metadata_str)

            updated_offset = await transaction.eventoffset.update(
                where={'id': 0},
                data={
                    "gallery_excuted_offset": int(seqno)
                }
            )
            if updated_offset == None:
                raise Exception(
                    f'[Curator Gallery]: Failed to update offset')

            new_state.new_offset.gallery_excuted_offset = updated_offset.gallery_excuted_offset
            return new_state, True
